### ISSUE TITLE
using TorchRay and extremal_perturbation method for custom datasets

### ISSUE THREAD

**Issue by mrdupadupa** [facebookresearch/TorchRay#14]
Dear TorchRay team,

Is it possible to use Torchray and especially extremal_perturbation method for the custom data sets? 
I didn't find any mentions in the documentation of how it could be used for other datasets not only for the ImageNet. In the `extremal_perturbation` function as the target could be used only `int` 0...999. I have tried to go through the source code of the `class Perturbation` for sources of the ImageNet classes or methods to change it for my classes however I didn't succeed. Could you help me and give an example of the code of how it could be done?

**Comment by ruthcfong** [comment]
Hi @mrdupadupa, thanks for your question and interest in extremal perturbations.

The `reward_func` parameter to the `extremal_perturbation` function handles setting the target class. By default, it uses `simple_reward` [function](https://github.com/facebookresearch/TorchRay/blob/master/torchray/attribution/extremal_perturbation.py#L224-L268), which expects an integer `target` (it's not restricted to be between 0-999 for ImageNet).

If you have a custom model that is similar to an image classifier --- i.e. the output is typically a 2D tensor -- batch x number of classes and the input is typically a 4D tensor -- batch x color channels x height x width), then you don't need to modify the code at all.

Depending on your dataset, you might want to play around with the following parameters to the `extremal_perturbation` function:
- `step` (int; default: 7): the size of a single mask unit in pixels (you might want to decrease step if you're working with fine-grain image data),
- `smooth` (int; default: 20): the amount of smoothing applied to an unsmoothed mask unit (if you decrease, the mask units will use less smoothing).
- `areas` (list; default: [0.1]): you might want to change / add more areas to the list.

If you want to use a model that has a different kind of output, then you would need to create your own reward function --- this can be done by adapting `simple_reward`.

Let us know if you have any other questions.

**Comment by mrdupadupa** [comment]
Hi @ruthcfong , thank you for your answer. It gave some clues for me. However, for me, it is still nor clear how could I use `extremal_perturbation` for my dataset. I prepared a minimal example (see the full code at the bottom). It is a simple image classifier with a greyscaled CIFAR10 and resnet18. My real code is similar. 

```
x = tensor1         #torch.Size([1, 1, 224, 224])
model = net.to(device)
x = x.to(device)

masks_1 = extremal_perturbation(
    model = model, input = x, target = 0,                                                   #what should I use here as a target? 
    reward_func=simple_reward,
    debug=True,
    areas=[0.1],
)
```
The target:  e.g 0 here. What is it actually? Is it a class from the ImageNet or is it a target channel from the simple reward function or is it my cifar class? 

I thought that the targets are different in the` simple_reward` and `contrastive_reward` functions. I suppose in the first one is a colour channel and for the second one is a class from ImageNet. But I could change it from 0 to 999 for my grayscale image(1 channel) from cifar with 10 classes. I'm using simple reward it should only 0.

The result doesn't look right for me for majority of values used in the` target`. And only when I'm using the number of class from the ImageNet similar to my images it gives more or less right area. It makes me think that I'm using ImageNet numbers. And I need to rewrite `simple_reward`

How could I use the `extremal_perturbation` for my case image classifier with a greyscaled CIFAR10 and resnet18 (see the code) and use it for my classes without linking them with ImageNet. Could you write an example?

This is the code:

```
import torch
import torchvision
import torchvision.transforms as transforms
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision.transforms import ToTensor
from PIL import Image
from torchray.attribution.extremal_perturbation import extremal_perturbation, contrastive_reward, simple_reward

#define transformations, loaders and NN

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

transform = transforms.Compose(
	[transforms.Grayscale(num_output_channels=1),
	transforms.ToTensor(),
	transforms.Normalize((0.5), (0.5))])

trainset = torchvision.datasets.CIFAR10(root='./data', train=True,
                                        download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=4,
                                          shuffle=True, num_workers=2)

testset = torchvision.datasets.CIFAR10(root='./data', train=False,
                                       download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=4,
                                         shuffle=False, num_workers=2)

classes = ('plane', 'car', 'bird', 'cat',
           'deer', 'dog', 'frog', 'horse', 'ship', 'truck')


__all__ = ['ResNet', 'resnet18']


def conv3x3(in_planes, out_planes, stride=1, groups=1, dilation=1):
    """3x3 convolution with padding"""
    return nn.Conv2d(in_planes, out_planes, kernel_size=3, stride=stride,
                     padding=dilation, groups=groups, bias=False, dilation=dilation)

def conv1x1(in_planes, out_planes, stride=1):
    """1x1 convolution"""
    return nn.Conv2d(in_planes, out_planes, kernel_size=1, stride=stride, bias=False)

class BasicBlock(nn.Module):
    expansion = 1

    def __init__(self, inplanes, planes, stride=1, downsample=None, groups=1,
                 base_width=64, dilation=1, norm_layer=None):
        super(BasicBlock, self).__init__()
        if norm_layer is None:
            norm_layer = nn.BatchNorm2d
        if groups != 1 or base_width != 64:
            raise ValueError('BasicBlock only supports groups=1 and base_width=64')
        if dilation > 1:
            raise NotImplementedError("Dilation > 1 not supported in BasicBlock")
        # Both self.conv1 and self.downsample layers downsample the input when stride != 1
        self.conv1 = conv3x3(inplanes, planes, stride)
        self.bn1 = norm_layer(planes)
        self.relu = nn.ReLU(inplace=True)
        self.conv2 = conv3x3(planes, planes)
        self.bn2 = norm_layer(planes)
        self.downsample = downsample
        self.stride = stride

        # Added another relu here
        self.relu2 = nn.ReLU(inplace=True)

    def forward(self, x):
        identity = x

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)

        if self.downsample is not None:
            identity = self.downsample(x)

        out += identity
        # Modified to use relu2
        out = self.relu2(out)

        return out

class Bottleneck(nn.Module):

    expansion = 4

    def __init__(self, inplanes, planes, stride=1, downsample=None, groups=1,
                 base_width=64, dilation=1, norm_layer=None):
        super(Bottleneck, self).__init__()
        if norm_layer is None:
            norm_layer = nn.BatchNorm2d
        width = int(planes * (base_width / 64.)) * groups
        # Both self.conv2 and self.downsample layers downsample the input when stride != 1
        self.conv1 = conv1x1(inplanes, width)
        self.bn1 = norm_layer(width)
        self.conv2 = conv3x3(width, width, stride, groups, dilation)
        self.bn2 = norm_layer(width)
        self.conv3 = conv1x1(width, planes * self.expansion)
        self.bn3 = norm_layer(planes * self.expansion)
        self.relu = nn.ReLU(inplace=True)
        self.downsample = downsample
        self.stride = stride

    def forward(self, x):
        identity = x

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)
        out = self.relu(out)
        out = self.conv3(out)
        out = self.bn3(out)

        if self.downsample is not None:
            identity = self.downsample(x)

        out += identity
        out = self.relu(out)
        return out
    
class ResNet(nn.Module):

    def __init__(self, block, layers, num_classes=1000, zero_init_residual=False,
                 groups=1, width_per_group=64, replace_stride_with_dilation=None,
                 norm_layer=None):
        super(ResNet, self).__init__()
        if norm_layer is None:
            norm_layer = nn.BatchNorm2d
        self._norm_layer = norm_layer

        self.inplanes = 64
        self.dilation = 1
        if replace_stride_with_dilation is None:
            # each element in the tuple indicates if we should replace
            # the 2x2 stride with a dilated convolution instead
            replace_stride_with_dilation = [False, False, False]
        if len(replace_stride_with_dilation) != 3:
            raise ValueError("replace_stride_with_dilation should be None "
                             "or a 3-element tuple, got {}".format(replace_stride_with_dilation))
        self.groups = groups
        self.base_width = width_per_group
        self.conv1 = nn.Conv2d(1, self.inplanes, kernel_size=7, stride=2, padding=3,
                               bias=False)
        self.bn1 = norm_layer(self.inplanes)
        self.relu = nn.ReLU(inplace=True)
        self.maxpool = nn.MaxPool2d(kernel_size=3, stride=2, padding=1)
        self.layer1 = self._make_layer(block, 64, layers[0])
        self.layer2 = self._make_layer(block, 128, layers[1], stride=2,
                                       dilate=replace_stride_with_dilation[0])
        self.layer3 = self._make_layer(block, 256, layers[2], stride=2,
                                       dilate=replace_stride_with_dilation[1])
        self.layer4 = self._make_layer(block, 512, layers[3], stride=2,
                                       dilate=replace_stride_with_dilation[2])
        self.avgpool = nn.AdaptiveAvgPool2d((1, 1))
        self.fc = nn.Linear(512 * block.expansion, num_classes)

        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
            elif isinstance(m, (nn.BatchNorm2d, nn.GroupNorm)):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)

        if zero_init_residual:
            for m in self.modules():
                if isinstance(m, Bottleneck):
                    nn.init.constant_(m.bn3.weight, 0)
                elif isinstance(m, BasicBlock):
                    nn.init.constant_(m.bn2.weight, 0)

    def _make_layer(self, block, planes, blocks, stride=1, dilate=False):
        norm_layer = self._norm_layer
        downsample = None
        previous_dilation = self.dilation
        if dilate:
            self.dilation *= stride
            stride = 1
        if stride != 1 or self.inplanes != planes * block.expansion:
            downsample = nn.Sequential(
                conv1x1(self.inplanes, planes * block.expansion, stride),
                norm_layer(planes * block.expansion),
            )

        layers = []
        layers.append(block(self.inplanes, planes, stride, downsample, self.groups,
                            self.base_width, previous_dilation, norm_layer))
        self.inplanes = planes * block.expansion
        for _ in range(1, blocks):
            layers.append(block(self.inplanes, planes, groups=self.groups,
                                base_width=self.base_width, dilation=self.dilation,
                                norm_layer=norm_layer))

        return nn.Sequential(*layers)

    def _forward_impl(self, x):
        # See note [TorchScript super()]
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu(x)
        x = self.maxpool(x)

        x = self.layer1(x)
        x = self.layer2(x)
        x = self.layer3(x)
        x = self.layer4(x)

        x = self.avgpool(x)
        x = torch.flatten(x, 1)
        x = self.fc(x)

        return x

    def forward(self, x):
        return self._forward_impl(x)

def _resnet(block, layers, pretrained, progress, **kwargs):
    model = ResNet(block, layers, **kwargs)
    if pretrained:
        state_dict = load_state_dict_from_url(model_urls[arch],
                                              progress=progress)
        
        #model.load_state_dict(state_dict)
    return model

def resnet18(pretrained=False, progress=True, **kwargs):

    return _resnet(BasicBlock, [2, 2, 2, 2], pretrained, progress,
                   **kwargs)

net = resnet18()
net = net.to(device)

criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)

#train NN

for epoch in range(1):  # loop over the dataset multiple times

    running_loss = 0.0
    for i, data in enumerate(trainloader, 0):
        # get the inputs; data is a list of [inputs, labels]
        inputs, labels = data[0].to(device), data[1].to(device)

        # zero the parameter gradients
        optimizer.zero_grad()

        # forward + backward + optimize
        outputs = net(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

        # print statistics
        running_loss += loss.item()
        if i % 2000 == 1999:    # print every 2000 mini-batches
            print('[%d, %5d] loss: %.3f' %
                  (epoch + 1, i + 1, running_loss / 2000))
            running_loss = 0.0

print('Finished Training')

#prepare the image 

PATH = './cifar_net.pth'
torch.save(net.state_dict(), PATH)

ind1 = 3

dataiter = iter(testloader)
images, labels = dataiter.next()

input = images[ind1].unsqueeze(0)
input.requires_grad = True

img = torchvision.utils.make_grid(images[ind1])
img = img.mean(dim=0)
img = (img * 0.5) + 0.5     # unnormalize
npimg = img.numpy()
matplotlib.image.imsave('name.png', npimg, cmap="Greys")

sourceimage= Image.open("name.png")   # original image 
resized_image_P = sourceimage.resize((224, 224), resample=Image.BILINEAR)  # resized image of size 224x224
resized_image = resized_image_P.convert("L") 
resized_image.save("name.png")

resized_image_data = ToTensor()(resized_image).cpu().unsqueeze(0)
resized_image_data.requires_grad = True
tensor1 = resized_image_data

x = tensor1

model = net.to(device)
x = x.to(device)

masks_1 = extremal_perturbation(
    model, x, 0,
    reward_func=simple_reward,
    debug=True,
    areas=[0.1],
    
)
```

**Comment by giangnguyen2412** [comment]
If you are looking for script to run pointing game on ImageNet, you can reach me!
